pymc/sampling/forward.py

```diff
@@ -497,22 +499,100 @@ def sample_posterior_predictive(
 
-        thinned_idata = idata.sel(draw=slice(None, None, 5))
-        with model:
-            idata.extend(pymc.sample_posterior_predictive(thinned_idata))
+        from logging import getLogger
+        import pymc as pm
+
+        # Some environments like google colab suppress the default logging output of PyMC
+        getLogger("pymc").setLevel("INFO")
+
+        kwargs = {"progressbar": False, "random_seed": 0}
+
+        with pm.Model() as model:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          z = pm.Normal("z", x + y**2)
+          det = pm.Deterministic("det", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
+
+          idata = pm.sample(tune=10, draws=10, chains=2, **kwargs)
+
+          # Default behavior
+          # Posterior predictive draws of `obs`, conditioned on the posterior draws of `z`
+          pm.sample_posterior_predictive(idata, var_names=["obs"], **kwargs)
+          # Sampling: [obs]
+
+          # Deterministic re-computation of `det` given the posterior draws of `z`
+          pm.sample_posterior_predictive(idata, var_names=["det"], **kwargs)
+          # Sampling: []
+
+          # Posterior predictive draws of `z` and `det`, conditioned on the posterior draws of `x` and `y`
+          # It does not matter that `z` was not an observed variable.
+          pm.sample_posterior_predictive(idata, var_names=["z", "det"], **kwargs)
+          # Sampling: [z]
 
-    Generate 5 posterior predictive samples per posterior sample.
+          # Posterior predictive draws of `y`, `z` and `det`, conditioned on the posterior draws of `x`
+          # Note that `y` samples are equivalent to the prior, since no other variables influence it.
+          pm.sample_posterior_predictive(idata, var_names=["y", "z", "det"], **kwargs)
+          # Sampling: [y, z]
+
+          # Same as before, except `z` is not stored in the returned trace.
+          # For computing `det` we still have to resample `z` as it depends on `y`, which is also being sampled.
+          pm.sample_posterior_predictive(idata, var_names=["y", "det"], **kwargs)
+          # Sampling: [y, z]
+
+          # Everything is sampled, which is equivalent to calling `sample_prior_predictive`
+          pm.sample_posterior_predictive(idata, var_names=["x", "y", "z", "det"], **kwargs)
+          # Sampling: [x, y, z]
+
+
+        # Note that "sampling" a dependent deterministic doesn't force random variables to be sampled.
+        with pm.Model() as model:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          det_xy = pm.Deterministic("det_xy", x + y**2)
+          z = pm.Normal("z", x + y**2)
+          det_z = pm.Deterministic("det_z", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
+
+          idata = pm.sample(tune=10, draws=10, chains=2, **kwargs)
+
+          pm.sample_posterior_predictive(idata, var_names=["det_xy", det_z"], **kwargs)
+          Sampling: [z]
```

Comment:
You said it doesn't force resample, but it's resampling?

Reply:
See my comment below! I was expecting this to be the behavior but it's not.

Reply:
This was just a typo, it was referring to a `det` in the previous example